Automated Smoketest: :x:
OS: centos 7.2.1511 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 54de1d761d1bfbb35848967dc7d08f6aabd65d2c
Number of Error(s): 4
/mnt/disk1/home/vamosax/smoketest/PR-11244/HPCC-Platform/ecl/hqlcpp/hqlcppds.cpp:3119:29: error: cannot declare variable \xe2\x80\x98rowsetAllocator\xe2\x80\x99 to be of abstract type \xe2\x80\x98EclccEngineRowAllocator\xe2\x80\x99

[Build error:](https://gist.github.com/51a32f8a2f1b495831f454f047dce511)
Build log [Build log](https://gist.github.com/HPCCSmoketest/51a32f8a2f1b495831f454f047dce511#file-result-18-05-29-07-37-25-log)

--- a/common/thorhelper/roxierow.cpp
+++ b/common/thorhelper/roxierow.cpp
@@ -280,24 +280,32 @@ public:
         heap.setown(rowManager.createFixedRowHeap(meta.getFixedSize()+CHECKER::extraSize, allocatorId | ACTIVITY_FLAG_ISREGISTERED | CHECKER::allocatorCheckFlag, (roxiemem::RoxieHeapFlags)flags));
     }
 
-    virtual void * createRow()
+    virtual void * createRow() override
     {
         return heap->allocate();
     }
 
-    virtual void * createRow(size32_t & allocatedSize)
+    virtual void * createRow(size32_t & allocatedSize) override
     {
         allocatedSize = meta.getFixedSize();
         return heap->allocate();
     }
 
-    virtual void * resizeRow(size32_t newSize, void * row, size32_t & size)
+    virtual void * createRow(size32_t initialSize, size32_t & allocatedSize) override
+    {
+        size32_t fixedSize = meta.getFixedSize();
+        assertex(initialSize == fixedSize);
+        allocatedSize = fixedSize;
+        return heap->allocate();
+    }
+
+    virtual void * resizeRow(size32_t newSize, void * row, size32_t & size) override
     {
         throwUnexpected();
         return NULL;
     }
 
-    virtual void * finalizeRow(size32_t finalSize, void * row, size32_t oldSize)
+    virtual void * finalizeRow(size32_t finalSize, void * row, size32_t oldSize) override
     {
         if (!meta.needsDestruct() && !CHECKER::allocatorCheckFlag)
             return row;
@@ -332,26 +340,24 @@ public:
         heap.setown(rowManager.createVariableRowHeap(allocatorId | ACTIVITY_FLAG_ISREGISTERED | CHECKER::allocatorCheckFlag, (roxiemem::RoxieHeapFlags)flags));
     }
 
-    virtual void * createRow()
+    virtual void * createRow() override
     {
         memsize_t allocSize = meta.getInitialSize();
         memsize_t capacity;
         return heap->allocate(allocSize+CHECKER::extraSize, capacity);
     }
 
-    virtual void * createRow(size32_t & allocatedSize)
+    virtual void * createRow(size32_t & allocatedSize) override
     {
-        const memsize_t allocSize = meta.getInitialSize();
-        memsize_t newCapacity; // always initialised by allocate
-        void * row = heap->allocate(allocSize+CHECKER::extraSize, newCapacity);
-        //This test should get constant folded to avoid the decrement when not checked.
-        if (CHECKER::extraSize)
-            newCapacity -= CHECKER::extraSize;
-        allocatedSize = (size32_t)newCapacity;
-        return row;
+        return doCreateRow(meta.getInitialSize(), allocatedSize);
     }
 
-    virtual void * resizeRow(size32_t newSize, void * row, size32_t & size)
+    virtual void * createRow(size32_t initialSize, size32_t & allocatedSize) override
+    {
+        return doCreateRow(initialSize, allocatedSize);
+    }
+
+    virtual void * resizeRow(size32_t newSize, void * row, size32_t & size) override
     {
         const size32_t oldsize = size;  // don't need to include the extra checking bytes
         memsize_t newCapacity; // always initialised by resizeRow
@@ -362,7 +368,7 @@ public:
         return newrow;
     }
 
-    virtual void * finalizeRow(size32_t finalSize, void * row, size32_t oldSize)
+    virtual void * finalizeRow(size32_t finalSize, void * row, size32_t oldSize) override
     {
         if (!meta.needsDestruct() && !CHECKER::allocatorCheckFlag)
             return row;
@@ -381,6 +387,19 @@ public:
         //It is not legal to call releaseAllRows on a variable size allocator - they are not allocated in a single heap
         throwUnexpected();
     }
+
+protected:
+    void * doCreateRow(size32_t initialSize, size32_t & allocatedSize)
+    {
+        memsize_t newCapacity; // always initialised by allocate
+        void * row = heap->allocate(initialSize + CHECKER::extraSize, newCapacity);
+        //This test should get constant folded to avoid the decrement when not checked.
+        if (CHECKER::extraSize)
+            newCapacity -= CHECKER::extraSize;
+        allocatedSize = (size32_t)newCapacity;
+        return row;
+    }
+
 
 protected:
     Owned<roxiemem::IVariableRowHeap> heap;

--- a/ecl/hqlcpp/hqlcppds.cpp
+++ b/ecl/hqlcpp/hqlcppds.cpp
@@ -3022,6 +3022,7 @@ public:
 
 //Used for dynamically sizing rows.
     virtual void * createRow(size32_t & allocatedSize) override { throwUnexpected(); }
+    virtual void * createRow(size32_t initialSize, size32_t & allocatedSize) override { throwUnexpected(); }
     virtual void * resizeRow(size32_t newSize, void * row, size32_t & size) override { throwUnexpected(); }
     virtual void * finalizeRow(size32_t newSize, void * row, size32_t oldSize) override { throwUnexpected(); }
 

--- a/rtl/eclrtl/rtlds.cpp
+++ b/rtl/eclrtl/rtlds.cpp
@@ -275,9 +275,8 @@ byte * RtlDynamicRowBuilder::ensureCapacity(size32_t required, const char * fiel
     if (required > maxLength)
     {
         if (!self)
-            create();
-
-        if (required > maxLength)
+            self = (byte *)rowAllocator->createRow(required, maxLength);
+        else if (required > maxLength)
         {
             void * next = rowAllocator->resizeRow(required, self, maxLength);
             if (!next)

--- a/rtl/include/eclhelper.hpp
+++ b/rtl/include/eclhelper.hpp
@@ -44,8 +44,8 @@ typedef unsigned short UChar;
 
 //Should be incremented whenever the virtuals in the context or a helper are changed, so
 //that a work unit can't be rerun.  Try as hard as possible to retain compatibility.
-#define ACTIVITY_INTERFACE_VERSION      699
-#define MIN_ACTIVITY_INTERFACE_VERSION  699             //minimum value that is compatible with current interface
+#define ACTIVITY_INTERFACE_VERSION      650
+#define MIN_ACTIVITY_INTERFACE_VERSION  650             //minimum value that is compatible with current interface
 
 typedef unsigned char byte;
 
@@ -259,6 +259,7 @@ interface IEngineRowAllocator : extends IInterface
 
 //Used for dynamically sizing rows.
     virtual void * createRow(size32_t & allocatedSize) = 0;
+    virtual void * createRow(size32_t initialSize, size32_t & allocatedSize) = 0;
     virtual void * resizeRow(size32_t newSize, void * row, size32_t & size) = 0;            //NB: size is updated with the new size
     virtual void * finalizeRow(size32_t newSize, void * row, size32_t oldSize) = 0;
 

